node/src/test_utils/recorder_stop_conditions.rs

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) 2019, MASQ (https://masq.ai) and/or its affiliates. All rights reserved.
+
+#![cfg(test)]
+
+use std::any::{Any, TypeId};
+
+pub enum StopConditions {
+    Single(StopCondition),
+    Any(Vec<StopCondition>),
+    All(Vec<StopCondition>),
+}
+
+pub enum StopCondition {
+    StopOnType(TypeId),
+    StopOnMatch {
+        exemplar: BoxedMsgExpected,
+    },
+    StopOnPredicate {
+        predicate: Box<dyn Fn(RefMsgExpected) -> bool + Send>,
+    },
+}
+
+pub type BoxedMsgExpected = Box<dyn Any + Send>;
+pub type RefMsgExpected<'a> = &'a (dyn Any + Send);
+
+impl StopConditions {
+    pub fn resolve_stop_conditions<T: PartialEq + Send + 'static>(&mut self, msg: &T) -> bool {
+        if let Some(matched) = self.inspect_immutable::<T>(msg) {
+            matched
+        } else {
+            self.inspect_mutable::<T>(msg)
+        }
+    }
+
+    fn inspect_immutable<T: PartialEq + Send + 'static>(&self, msg: &T) -> Option<bool> {
+        match self {
+            StopConditions::Single(stop_condition) => {
+                Some(stop_condition.resolve_condition::<T>(msg))
+            }
+            StopConditions::Any(stop_conditions) => {
+                Some(Self::resolve_any::<T>(stop_conditions, msg))
+            }
+            StopConditions::All(_) => None,
+        }
+    }
+
+    fn resolve_any<T: PartialEq + Send + 'static>(
+        conditions: &Vec<StopCondition>,
+        msg: &T,
+    ) -> bool {
+        conditions
+            .iter()
+            .any(|condition| condition.resolve_condition::<T>(msg))
+    }
+
+    fn inspect_mutable<T: PartialEq + Send + 'static>(&mut self, msg: &T) -> bool {
+        match self {
+            StopConditions::All(conditions) => {
+                let indexes_to_remove = Self::indexes_of_matched_conditions(conditions, msg);
+                Self::remove_matched_conditions(conditions, indexes_to_remove);
+                conditions.is_empty()
+            }
+            _ => unreachable!("something is wrong"),
```

Comment:
This message can probably be improved. Imagine that somebody else quoted it to you and you had to track it down.